documentation/writing_tasks.md

```diff
@@ -1077,3 +1077,453 @@ path.
+> establish connections with the socket opened by `pry-remote`, which may be
+> difficult if your target restricts incoming connections.
+
+You can quickly install the `pry-remote` gem using Bolt's Ruby:
+
+```shell
+$ /opt/puppetlabs/bolt/bin/gem install --user-install pry-remote
+```
+
+If you are running Bolt on Windows with the default install location, run
+the following command instead:
+
+```powershell
+> "C:/Program Files/Puppet Labs/Bolt/bin/gem.bat" install --user-install pry-remote
+```
+
+Then, add the following line to your task wherever you want to begin debugging:
+
+```ruby
+require 'pry-remote'; binding.remote_pry
+```
+
+You can then open a connection to a target using the `pry-remote` command:
+
+```shell
+$ pry-remote
+
+Frame number: 0/4
+
+From: /tmp/4f9dcfa3-ce0c-49e2-bcf5-8d761b202186/ruby/tasks/init.rb @ line 9 MyClass#task:
+
+     6: def task(opts)
+ =>  9:   require 'pry-remote'; binding.remote_pry
+    10:
+```
+
+📖 **Related information**
+- [`pry-remote` documentation](https://www.rubydoc.info/gems/pry-remote)
+- [`pry` documentation](http://pry.github.io/)
+
+#### PowerShell tasks
+
+PowerShell tasks can take advantage of the `Set-PSBreakpoint` cmdlet to debug
+tasks that are run as scripts. The `Set-PSBreakpoint` cmdlet can set a
+breakpoint at a specific line of your task, pausing execution of the task
+so you can examine the code and step through it line by line.
+
+To set a breakpoint at a specific line of your task, run the `Set-PSBreakpoint`
+cmdlet:
+
+```powershell
+> Set-PSBreakpoint -Script mytask.ps1 -Line <line number>
+```
+
+You can then run the task as a script. Execution of the task will pause at
+the breakpoint.
+
+```powershell
+> ./mytask.ps1
+```
+
+📖 **Related information**
+- [`Set-PSBreakpoint`
+  documentation](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/set-psbreakpoint?view=powershell-7)
+
+### Task helper methods
+
+Both the [Python task
+helper](https://github.com/puppetlabs/puppetlabs-python_task_helper) and [Ruby
+task helper](https://github.com/puppetlabs/puppetlabs-ruby_task_helper) include
+methods that can help you debug a task. The `debug` method logs arbitrary values
+as debugging messages, while the `debug_statements` method returns an array of
+the logged debugging messages.
+
+When a task using a task helper library raises a `TaskError`, the error will
+include any logged debugging messages under the `details` key. You can also add
+debugging statements when raising a `TaskError` yourself by calling
+`debug_statements` and adding the result under the `details` key.
+
+> 🔩 **Tip:** When running a task with `bolt task run`, use the `--format json`
+> option to see the full result from the task, including the value of the
+> `details` key.
+
+#### Python tasks
+
+The following Python task includes a few debugging statements which describe
+what the task is doing and the results from a couple arithmetic operations:
```

Comment:
We should also include the `task.json` for this so that the helper files load correctly.